hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java

```diff
@@ -136,7 +137,14 @@ public CompletableFuture<Message> applyTransaction(
     try {
       final SCMRatisRequest request = SCMRatisRequest.decode(
           Message.valueOf(trx.getStateMachineLogEntry().getLogData()));
-      applyTransactionFuture.complete(process(request));
+
+      try {
+        applyTransactionFuture.complete(process(request));
+      } catch (SCMException ex) {
+        // SCMException is considered as logical rejection and is returned to
+        // Ratis client.
```

Comment:
From `SCMException#ResultCodes`, `IO_EXCEPTION` and `INTERNAL_ERROR` should probably still crash the SCM

Reply:
Good point, updated.